virtcontainers/pod.go

```diff
@@ -378,9 +378,9 @@ func (podConfig *PodConfig) valid() bool {
 		return false
 	}
 
-	if _, err := newHypervisor(podConfig.HypervisorType); err != nil {
-		podConfig.HypervisorType = QemuHypervisor
-	}
+	// HACK: This forces the plugin to be used.
+	podConfig.HypervisorType = PluginHypervisor
+	podConfig.HypervisorConfig.PluginPath = "$GOPATH/src/github.com/kata-containers/runtime/plugins/hypervisor/qemu/qemu.so"
```

Comment:
Can we have a concept of in-tree plugins that still works in a plugin option manner but we do not load them by dynamic libraries?

I thought we wanted that in the meeting and confirmed with @sameo in his slide (https://docs.google.com/presentation/d/1dbGrD1h9cpuqAPooiEgtiwWDGCYhVPdatq7owsKHDEQ/edit#slide=id.g35183e1ff3_0_25)

Reply:
Agreed. This is starting to sound like the kernel ;)

I've just had a play around a little with this concept and have some very crude code to handle both static and dynamic registration using something like this:

```
type PluginHandler interface {
    Foo(rope string) error
    Bar(mad bool) error
    Baz(snowballs float64) error
}


// PluginDetails allows a package to register itself as either a:                              
//                                                                                             
// - static plugin (sets Handler)                                                              
// - dynamic plugin (Sets Dynamic+Path)                                                        
type PluginDetails struct {                                                                    
        // Name of the plugin                                                                  
        Name string                                                                            
                                                                                               
        // set if plugin is dynamically loadable                                               
        Dynamic bool                                                                           
                                                                                               
        // Full path to shared object containing dynamically-loadable module                   
        Path string                                                                            
                                                                                               
        // set if Dynamic is false. Used for handling statically registered                    
        // modules.                                                                            
        Handler PluginHandler                                                                
}

// called from `main` for all modules to use (static and dynamic)
func RegisterPlugin(details PluginDetails) (PluginHandler, error)
```

There is one thing worth noting: the plugin package *only* loads code into the `main` package (although the dynamic loading can itself be performed in another package). This is mentioned (all too briefly) on https://golang.org/pkg/plugin/. fwics, this means we'll need something like:

  ```
  plugins/stub_foo.go

    package main
    import "pkg/foo"

  plugins/stub_bar.go

      package main
      import "pkg/bar"

  pkg/foo/foo.go

    package foo

  pkg/bar/bar.go

    package bar

  ```
  ... and the `main` code will then call (probably through another package):
  ```
  plugin.Open("stub_foo.so")
  plugin.Open("stub_bar.so")
  ```

So, code in the `main` package can only load new code into the `main` package. Hence, to dynamically load a new package, you need to create a stub that is part of `main` and have *that* `import` the real package you want to load.